File: src/engine/Scene.jl
module SceneModule
    using ..JulGame
    
    export Scene
    mutable struct Scene
        camera::Union{Nothing, JulGame.CameraModule.Camera}
        colliders::Vector{Any}
        entities::Vector{Any}
        rigidbodies::Vector{Any}
        uiElements::Vector{Any}
        name::String
        batchedLayers::Dict{Int, Any}  # Static sprite batching: layer => BatchedLayer

        function Scene()
            this = new()

            this.camera = nothing
            this.colliders = []
            this.entities = []
            this.rigidbodies = []
            this.uiElements = []
            this.batchedLayers = Dict{Int, Any}()

            return this
        end
    end

    function get_entity_by_name(this::Scene, name)
        for entity in this.entities
            if entity.name == name
                return entity
            end
        end

        @debug "No entity with name $name found"
        return nothing
    end

    function get_entity_by_name(name)
        return get_entity_by_name(MAIN.scene, name)
    end

    function get_entities_by_name(this::Scene, name)
        entities = []
        for entity in this.entities
            if entity.name == name
                push!(entities, entity)
            end
        end

        if length(entities) == 0
            @debug "No entity with name $name found"
        end
        return entities
    end

    function get_entities_by_name(name)
        return get_entities_by_name(MAIN.scene, name)
    end

    function get_entity_by_id(this::Scene, id)
        for entity in this.entities
            if entity.id == id
                return entity
            end
        end

        @debug "No entity with id $id found"
        return nothing
    end

    function get_entity_by_id(id::String)
        return get_entity_by_id(MAIN.scene, id)
    end

    function get_ui_element_by_name(this::Scene, name)
        for entity in this.uiElements
            if entity.name == name
                return entity
            end
        end

        @debug "No entity with name $name found"
        return nothing
    end

    function get_ui_element_by_name(name)
        return get_ui_element_by_name(MAIN.scene, name)
    end

    function get_ui_element_by_id(this::Scene, id)
        for element in this.uiElements
            if element.id == id
                return element
            end
        end

        @debug "No ui element with id $id found"
        return nothing
    end

    function get_ui_element_by_id(id::String)
        return get_ui_element_by_id(MAIN.scene, id)
    end
end

File: ts/_generated/src/engine/Scene.ts
export {}
import { Camera } from "./Camera/Camera";


    // using ..JulGame
    
    
    class Scene {
        camera: null | Camera
        colliders: any[]
        entities: any[]
        rigidbodies: any[]
        uiElements: any[]
        name: string
        batchedLayers: Record<number, any>  // Static sprite batching: layer => BatchedLayer

        constructor() {
            

            this.camera = null
            this.colliders = []
            this.entities = []
            this.rigidbodies = []
            this.uiElements = []
            this.batchedLayers = {}

        }
    }

    function get_entity_by_name(selfOrName: Scene | string, name?: string) {
        if (typeof selfOrName === "string") {
            return get_entity_by_name(MAIN.scene, selfOrName)
        }
        const self = selfOrName
        for (const entity of self.entities) {
            if (entity.name == name) {
                return entity
            }
        }

        console.debug(`No entity with name ${name} found`)
        return null
    }

    function get_entities_by_name(selfOrName: Scene | string, name?: string) {
        if (typeof selfOrName === "string") {
            return get_entities_by_name(MAIN.scene, selfOrName)
        }
        const self = selfOrName
        let entities = []
        for (const entity of self.entities) {
            if (entity.name == name) {
                entities.push(entity)
            }
        }

        if (entities.length == 0) {
            console.debug(`No entity with name ${name} found`)
        }
        return entities
    }

    function get_entity_by_id(selfOrName: Scene | string, id?: string) {
        if (typeof selfOrName === "string") {
            return get_entity_by_id(MAIN.scene, selfOrName)
        }
        const self = selfOrName
        for (const entity of self.entities) {
            if (entity.id == id) {
                return entity
            }
        }

        console.debug(`No entity with id ${id} found`)
        return null
    }

    function get_ui_element_by_name(selfOrName: Scene | string, name?: string) {
        if (typeof selfOrName === "string") {
            return get_ui_element_by_name(MAIN.scene, selfOrName)
        }
        const self = selfOrName
        for (const entity of self.uiElements) {
            if (entity.name == name) {
                return entity
            }
        }

        console.debug(`No entity with name ${name} found`)
        return null
    }

    function get_ui_element_by_id(selfOrName: Scene | string, id?: string) {
        if (typeof selfOrName === "string") {
            return get_ui_element_by_id(MAIN.scene, selfOrName)
        }
        const self = selfOrName
        for (const element of self.uiElements) {
            if (element.id == id) {
                return element
            }
        }

        console.debug(`No ui element with id ${id} found`)
        return null
    }
export { Scene, get_entities_by_name, get_entity_by_id, get_entity_by_name, get_ui_element_by_id, get_ui_element_by_name }
